File: src_python/data_loader.py
class Transaction:
    def __init__(self, items):
        self.items = list(items)


class DataLoader:
    def __init__(self, file_path: str):
        self.file_path = file_path
        self.transactions = {}

    def __parse_items(self, line: str):
        clean_line = line.strip()
        if not clean_line:
            return []

        if "{" in clean_line and "}" in clean_line:
            clean_line = clean_line[clean_line.find("{") + 1 : clean_line.rfind("}")]

        return [item.strip() for item in clean_line.split(",") if item.strip()]

    def load_transactions(self):
        with open(self.file_path, "r", encoding="utf-8") as f:
            transaction_idx = 1
            for line in f:
                items = self.__parse_items(line)
                if not items:
                    continue

                transaction = Transaction(items)
                self.transactions[f"t{transaction_idx}"] = transaction
                transaction_idx += 1

    def get_transactions(self):
        if not self.transactions:
            self.load_transactions()

        return self.transactions

    def get_unique_items(self):
        unique_items = set()
        for transaction in self.transactions.values():
            unique_items.update(transaction.items)
        return unique_items

File: src/data_loader.jl
function parse_items(line::AbstractString)::Vector{String}
    clean_line = strip(line)
    isempty(clean_line) && return String[]

    if occursin("{", clean_line) && occursin("}", clean_line)
        start_idx = findfirst('{', clean_line)
        end_idx = findlast('}', clean_line)
        clean_line = clean_line[nextind(clean_line, start_idx):prevind(clean_line, end_idx)]
    end

    if occursin(",", clean_line)
        return [strip(item) for item in split(clean_line, ",") if !isempty(strip(item))]
    end

    return String.(split(clean_line))
end

function load_transactions(loader::DataLoader)::Dict{String, Transaction}
    if !isempty(loader.transactions)
        return loader.transactions
    end

    transaction_idx = 1
    open(loader.file_path, "r") do io
        for line in eachline(io)
            items = parse_items(line)
            isempty(items) && continue

            loader.transactions["t$(transaction_idx)"] = Transaction(items)
            transaction_idx += 1
        end
    end

    return loader.transactions
end

load_transactions(file_path::AbstractString)::Dict{String, Transaction} =
    load_transactions(DataLoader(file_path))

function load_unique_items(loader::DataLoader)::Set{String}
    transactions = load_transactions(loader)
    unique_items = Set{String}()

    for transaction in values(transactions)
        union!(unique_items, transaction.items)
    end

    return unique_items
end
